include/irr/asset/CTexturePacker.h

```diff
@@ -53,20 +55,119 @@ class CSquareAddressAllocator
+        m_addrAlctr(14u, _tilesPerDim, _numLayers)
+    {
+        ICPUImage::SCreationParams params;
+        params.arrayLayers = _numLayers;
+        params.extent = {_tilesPerDim,_tilesPerDim,1u};
+        params.flags = static_cast<ICPUImage::E_CREATE_FLAGS>(0);
+        params.format = EF_R32_UINT;
+        params.mipLevels = core::findMSB(_tilesPerDim)+1;
+        params.samples = ICPUImage::ESCF_1_BIT;
+        params.type = IImage::ET_2D;
+        m_pageTable = ICPUImage::create(std::move(params));
+        
+        auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(m_pageTable->getCreationParameters().mipLevels);
+
+        uint32_t bufOffset = 0u;
+        for (uint32_t i = 0u; i < m_pageTable->getCreationParameters().mipLevels; ++i)
+        {
+            const uint32_t tilesPerLodDim = _tilesPerDim>>i;
+            const uint32_t regionSz = tilesPerLodDim*tilesPerLodDim*_numLayers*4u;
+            auto& region = (*regions)[i];
+            region.bufferOffset = bufOffset;
+            region.bufferImageHeight = 0u;
+            region.bufferRowLength = tilesPerLodDim*4u;
+            region.imageExtent = {tilesPerLodDim,tilesPerLodDim,1u};
+            region.imageOffset = {0,0,0};
+            region.imageSubresource.baseArrayLayer = 0u;
+            region.imageSubresource.layerCount = _numLayers;
+            region.imageSubresource.mipLevel = i;
+            region.imageSubresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+
+            bufOffset += regionSz;
+        }
+        auto buf = core::make_smart_refctd_ptr<ICPUBuffer>(bufOffset);
+
+        m_pageTable->setBufferAndRegions(std::move(buf), regions);
+    }
+
+    core::smart_refctd_dynamic_array<STexOffset> alloc(const ICPUImage* _img, const ICPUImage::SSubresourceRange& _subres)
+    {
+        auto extent = _img->getCreationParameters().extent;
+        uint32_t levelCount = 1u;
+        for (uint32_t i = 0u; (extent.width>>(_subres.baseMipLevel+i) > m_pgSzxy) || (extent.height>>(_subres.baseMipLevel+i) > m_pgSzxy); ++i)
+            levelCount = i;
+        levelCount = std::min(_subres.levelCount, levelCount);
+
+        auto offsets = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<STexOffset>>(levelCount);
+        for (uint32_t i = 0u; i < levelCount; ++i)
+        {
+            const uint32_t w = std::max(extent.width>>(_subres.baseMipLevel+i),1u);
+            const uint32_t h = std::max(extent.height>>(_subres.baseMipLevel+i),1u);
+            const uint32_t pgSqrSz = core::roundUpToPoT((std::max(w,h)+m_pgSzxy-1u)/m_pgSzxy);
+
+            const uint32_t addr = m_addrAlctr.alloc_addr(pgSqrSz);
+            (*offsets)[i] = (addr==CSquareAddressAllocator<uint32_t>::invalid_address) ? 
+                STexOffset::invalid_addr :
+                STexOffset(m_addrAlctr.unpackAddress_x(addr), m_addrAlctr.unpackAddress_y(addr), m_addrAlctr.unpackAddress_layer(addr));
```

Comment:
you only allocate once (for the base mip-map level only) since the texture allocation will be PoT, so all other mip-map layers can be inferred.

Reply:
wouldnt then at least half of allocated space be unused?

Reply:
Well, no why? the page table is a regular texture.

If you allocate a texture which is (16,16) pages, then due to morton code + alignment it can only have an offset that is a multiple of its size ;)

hence possible offset you'd get allocated in the base mip would be (0/8/16/24/32/....,0/8/16/24/32/....)

The remaining mip level offsets you can infer by dividing th base leve offset by `exp2(mipLevel)` , because of magical alignment and the quadtree-like structure you will never suffer from integer truncation.

Obviously when you get to the 1x1 page tile level, you stop needing to worry about the remaining mip levels (we'll make sure the next 2 contiguous pages get allocated for a mip tail) just make sure you mark that page table entry with the `TAIL_BIT` flag (which still leaves you with enough bits to address 2*1024*1024*1024 separate physical pages, which is insane - or 32k if you use 16bit physical table pointers which would still allow you to keep 512 Megapixels resident in memory with a 128x128 page tile).

Reply:
ah, you're talking about page table. This code here is allocation of tiles into physical texture

Reply:
Square address allocator is for virtual address

PoolAddressAllocator with no alignment at all is for physical pages and the only value being saves should be a uint16_t ;)

Reply:
oh, allocation is done a lot different than you wanted it then XD Actually now i see more sense to it. Luckily it's no problem to change it